### PR TITLE
Grain Services

### PR DESCRIPTION
**Pull request by jamescarter-le**
I have attempted to implement as advised by Sergey the `GrainService` and `GrainServiceClient` types as described by his Option 1 in Issue #2444 
 
 The surface area of this change is large, because I have added a `IGrainService` type, analogues to `ISystemTarget`. I assumed `ISystemTarget` was to remain `internal`, and so I added GrainService as a direct clone of SystemTarget, except it implements LocalReminderService type RingRange knowledge.
 
 Some notes and assumptions made along the way:
  - `SystemTarget` was to remain internal, required change throughout code-base to deal with this additional base `Addressable` type. Changed:
    - `Silo`
    - `SchedulingContext`
    - `ResponseWorkItem`
    - `RequestWorkItem`
    - `IncomingMessageAgent`
    - `InsideRuntimeClient`
    - `ActivationDirectory`
    - `GrainReference`
    - `UniqueKey`
    - `ActivationId`
    - If we were to make `SystemTarget` public, we could prevent all these changes, and derive `GrainClient` directly from `SystemTarget`
 - `GrainServiceClient` added - `ReminderRegistry` is now derived from this and uses the `GrainService` it resolves through `InternalRuntimeClient`
 - `GrainClient` added
    - Basically a `SystemTarget` crossed with the Start, Stop, RingRange properties of `LocalReminderService`
    - Provides virtual methods for derived types to override.
    - Responsible for subscribing to the `IConsistentRingProvider`, keeping this type `internal`
 - `LocalReminderService` modified
    - Now derives from `GrainService` instead of `SystemTarget`
    - Removed all code moved to `GrainClient`
    - This type now solely handles Reminder specific responsibilities.
    - Uses the virtual methods provided in `GrainClient` to hook into the Start/Stopped flow and respond to Ring Range change events.
  
This change is by someone not fully aware of the total scope of Orleans, and will contain significant issues. It is implemented only to convey an idea of how to proceed to allow developers using Orleans to provide RingRange based services, such as those similar to the Reminder Service.

@sergeybykov - Please could you provide feedback on whether SystemTarget should remain internal or be exposed, meaning GrainService can derive from it and we remove the far reaching changes into the code base, however this would require making further types, GrainReference, UniqueKey and more public also.

### CODE REVIEW

**Comment by sergeybykov** [comment]
@jamescarter-le, you beat me by a couple of hours - I was prototyping the change for you - https://github.com/sergeybykov/orleans/tree/grainservice. :-)

I made `SystemTarget` and `ISystemTarget` public to minimize the amount of changes. Please take a look/

I also wanted to prototype an encapsulation option in addition to the inheritance one, to see if it would be cleaner.

**Comment by jamescarter-le** [comment]
@sergeybykov Looks like we both came up with the same in the end. I wasn't sure if you wanted to expose SystemTarget, I can make GrainService inherit from SystemTarget instead.

Would you like to implement the encapsulation option, and keep a private SystemTarget inside GrainService, exposed through ISystemTarget and ISystemTargetBase on GrainService?

Happy to continue to progress this item with your guidance.

**Comment by sergeybykov** [comment]
I looked at the encapsulation option briefly. Looks like it would be much more complicated because of the interface to class mapping that happens al load time, and would have to change to inject such mapping at run time to reflect encapsulation. I think it's not worth it.

I think it's better to expose `SystemTarget` and `ISytemTarget` as public at this point. We may decide to change the naming last minute, as somebody suggested on Gitter, but that would be trivial to do.

So, my suggestion is to take the inheritance route, as it's much less disruptive.

>Also - currently SystemTargets have Constant GrainIds - how would you envisage we deal with this when adding application specific ones?

It seems to me that we need to use type codes for such system targets, to differentiate between different types. I don't remember why they are not used today.

So `UniqueKey.NewSystemTargetKey` would have to change to `public static UniqueKey NewSystemTargetKey(short systemId, long typeCodeData=0);` 
and we'll need a new method `GrainId.GetGrainServiceGrainId(short id, TypeInfo grainServiceType)`, similar to `GetSystemTargetGrainId`, to construct a `GrainId` with the type code and a key.

We can start with creating a single `GrainService` activation of a type per silo, and use key 0 for them. We can later expand to support more than one activation per silo, and assign sequential keys to them: 1,2,3...

On the `GrainServiceClient` side we'll need to wire resolution of the `GrainService` type via the type code map, just like we do for grains. Or we can cheat initially by passing TypeInfo of the target `GrainService` type, and calculate its type code as part of `GrainFactory.GetSystemTarget` call.

This is what comes to my mind. I'm curious if people have simpler/better ideas.

**Comment by sergeybykov** [comment]
@jamescarter-le Let me know if I confused you too much. :-) We'd be happy to help here.

**Comment by jamescarter-le** [comment]
Hi @sergeybykov  - Not at all that is very helpful and I am following you.

I will revert my change and implement again but by changing the access modifier of SystemTarget to public instead of making GrainService distinct.

Happy to go with single grain services per Silo for now, with TypeCode as the differentiator in the GrainId.

Quick question, how are you intending to resolve these grain services inside a Grain?
Currently Reminders are helper methods on a Grain. We could add a GetGrainService<T(IGrainServiceClient)> call to a Grain, which queries the Runtime, however perhaps people would want to get these by DI? If we are expecting people to register them by DI, we could inject the GrainReference for the GrainService into the Grain expecting to consume the service. This should be okay as SystemTarget's cannot be called from Clients anyway according to what I can see.

I should be able to do more work on this over the weekend.

**Comment by sergeybykov** [comment]
>Quick question, how are you intending to resolve these grain services inside a Grain?
Currently Reminders are helper methods on a Grain. We could add a GetGrainService<T(IGrainServiceClient)> call to a Grain, which queries the Runtime, however perhaps people would want to get these by DI? If we are expecting people to register them by DI, we could inject the IGrainServiceClient into the Grain expecting to consume the service. This should be okay as SystemTarget's cannot be called from Clients (Silo Clients) anyway according to what I can see.

I think we don't want to expose this via a method on `Grain` class just yet. We can try adding a `GetGrainService<T>` to `IGrainFactory` first, and return a wrapper over a `GrainReference`.

We need a partitioning key to choose the target silo. We can start with implicitly using `CurrentActivationData.GrainReference` just like `GetReminderService` does today, and later consider if we also want to support other kinds of partitioning keys, for example, to allow clients to call grain services, too.

The reason we need a wrapper is that we'll need to recalculate the target silo for every request or upon any change to the cluster state because that will cause repartitioning of the system targets. `RegisterOrUpdateReminder` calls `GetReminderService` -> `MapGrainReferenceToSiloRing` -> `ConsistentRingProvider.GetPrimaryTargetSilo(hashCode)` on every call for that reason.

**Comment by jamescarter-le** [comment]
So if we return a GrainReference from GrainFactory, then we would be returning the IGrainService (SystemTarget) reference, as the GrainServiceClient is not IAddressable.

In this case - I'm not sure I follow what the GrainServiceClient should be doing - I thought that was its job, to address messages to the correct GrainService? See the GrainServiceClient.GrainService property:

https://github.com/dotnet/orleans/pull/2459/files#diff-3748abb74ef2a327f5a9e859d9ca19d2R15

I understand the to be addressed may move, and thought it was the GrainServiceClient's job to address the calls. Perhaps I've misunderstood.

File: src/OrleansRuntime/ReminderService/LocalReminderServiceFactory.cs [hunk 1]
```diff
@@ -17,14 +17,17 @@ public LocalReminderServiceFactory()
         internal IReminderService CreateReminderService(Silo silo, IGrainFactory grainFactory, TimeSpan iniTimeSpan)
         {
             var reminderServiceType = silo.GlobalConfig.ReminderServiceType;
-            logger.Info("Creating reminder system target for type={0}", Enum.GetName(typeof(GlobalConfiguration.ReminderServiceProviderType), reminderServiceType));
+            logger.Info("Creating reminder grain service for type={0}", Enum.GetName(typeof(GlobalConfiguration.ReminderServiceProviderType), reminderServiceType));
+
+            // At the point this is called the GrainClassData is not available for IReminderService - is this why Constants are used?
```
Comment (jamescarter-le) [inline src/OrleansRuntime/ReminderService/LocalReminderServiceFactory.cs]: Please advise how to get the TypeCode for the Grain.

File: src/OrleansRuntime/Silo/Silo.cs [hunk 2]
```diff
@@ -284,6 +286,8 @@ internal Silo(SiloInitializationParameters initializationParams)
                         services.TryAddSingleton<IConsistentRingProvider>(
                             sp => new ConsistentRingProvider(this.initializationParams.SiloAddress));
                     }
+
+                    services.TryAddSingleton<IGrainServiceClient<IReminderService>, GrainServiceClient<IReminderService>>();
```
Comment (jamescarter-le) [inline src/OrleansRuntime/Silo/Silo.cs]: Added to resolve the IReminderService from the InsideRuntimeClient in a common way.
When we add user defined IGrainServices we grab the GrainServiceClient from the DI.

File: test/Tester/GrainServiceTests/GrainServiceTests.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            var options = new TestClusterOptions(1);
+            options.ClusterConfiguration.UseStartupType<GrainServiceStartup>();
+
+            return new TestCluster(options);
+        }
+
+        [Fact, TestCategory("BVT"), TestCategory("Functional"), TestCategory("GrainServices")]
+        public async Task SimpleInvokeGrainService()
+        {
+            // We need to get the Silo to create the GrainService instances and register them as SystemTargets.
```
Comment (jamescarter-le) [inline test/Tester/GrainServiceTests/GrainServiceTests.cs]: I need to find a way to bootstrap creating the GrainService SystemTargets in the Silo on Startup.

File: src/OrleansRuntime/Services/GrainServiceClient.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        }
+        
+        /// <summary>
+        /// Resolves the Grain Reference invoking this request.
+        /// </summary>
+        protected GrainReference CallingGrainReference
+        {
+            get
+            {
+                return RuntimeClient.Current.CurrentActivationData.GrainReference;
```
Comment (ReubenBond) [inline src/OrleansRuntime/Services/GrainServiceClient.cs]: I'm trying to remove statics like this so that we can enable non-static grain clients

Comment (jamescarter-le) [inline src/OrleansRuntime/Services/GrainServiceClient.cs]: Thanks Reuben, It looks like IInternalGrainFactory and IRuntimeClient are passed into IServiceProvider so I can take an IServiceProvider in GrainServiceClient constructor as that is a public interface.

I will refactor this later and you can review.

**Comment by jamescarter-le** [comment]
I've added code to construct GrainServices in the Silo, but when I attempt to run the Tests I'm getting:

Exception thrown: 'Orleans.CodeGenerator.CodeGenerationException' in OrleansCodeGenerator.dll

Additional information: (1,18913): error CS0122: 'ITestHooksSystemTarget' is inaccessible due to its protection level

I wonder if someone wouldn't mind having a look at this for me - I can't see where I am attempting to expose anything internal?

**Comment by jamescarter-le** [comment]
Currently the Silo is constructing the Grain Service for each Service using Reflection to locate the constructor, supplied by Type names in the Config.

I'm sure you guys may have a better way of doing this, happy to take feedback.

File: test/Tester/GrainServiceTests/GrainServiceTests.cs (hunk 3 above)
Comment (jamescarter-le) [inline test/Tester/GrainServiceTests/GrainServiceTests.cs]: Done.

**Comment by sergeybykov** [comment]
I submitted a PR - https://github.com/jamescarter-le/orleans/pull/2 - with a few minor cleanup type of changes. Otherwise, looks good to me. I wonder what others think.

**Comment by sergeybykov** [comment]
I rebased your branch on top of current master and pushed the result to https://github.com/sergeybykov/orleans/tree/2459-3. Can you use that to update the PR branch?

We are looking to fix the two vNext build errors. I have trouble with the tooling. I asked @attilah to help.

**Comment by sergeybykov** [comment]
I pushed another commit to https://github.com/sergeybykov/orleans/tree/2459-3 to resolve the .NET Core build issue (thanks to @attilah).

Let us know if you are not sure how to integrate the rebased branch into yours.

**Comment by jamescarter-le** [comment]
Apologies I'm short on time at the moment - I will look to get the PR branch merged in shortly.

**Comment by sergeybykov** [comment]
@jamescarter-le If you don't have time, we could merge my rebased branch of this PR with all your commits instead. The commits are what matters in the end anyway. :-)

**Comment by jamescarter-le** [comment]
Hi Sergey, yes sorry happy to you to get the changes in however is easiest
for yourself.

Cheers, James

**Comment by ReubenBond** [comment]
Merged via #2531

**Pull request closed without merging**